pkg/agent/tcpconn/handler.go

```diff
@@ -0,0 +1,137 @@
+type NFQConfig struct {
+	// QueueID is an arbitrary integer used to identify a queue where a handler listens and a disruptor redirects target
+	// packets.
+	QueueID uint16
+	// RejectMark is an arbitrary integer which the handler uses to mark packets that need to be dropped.
+	RejectMark uint32
+}
+
+// DefaultNFQConfig contains default values for NFQ IDs. Only one disruptor/handler should be using this config at the
+// same time.
+var DefaultNFQConfig = NFQConfig{
+	QueueID:    1076,
+	RejectMark: 1076,
+}
+
+// RandomNFQConfig returns a NFQConfig with two random integers to be used as queue IDs and reject mark.
+func RandomNFQConfig() NFQConfig {
+	return NFQConfig{
+		QueueID:    uint16(rand.Int31()),
+		RejectMark: uint32(rand.Int31()),
+	}
+}
+
+// Handler is the interface implemented by objects that listen and handle packets in userspace.
+type Handler interface {
+	Handle(ctx context.Context) error
+}
+
+// NoopHandler is a no-op implementation of Handler.
+type NoopHandler struct{}
+
+func (nh NoopHandler) Handle(ctx context.Context) error {
+	<-ctx.Done()
+	return ctx.Err()
+}
+
+// NFQHandler implements the Handler interface using netfilter's nfqueue mechanism, reading packets sent to userspace by
+// netfilter and emits a veredict on whether a packet should be accepted or requeued.
+// NFQHandler in particular decides whether a connection should be terminated based on DropRate.
+// If a connection is decided to be terminated, it sets a flag on the packet and requeues it. Iptables rules, typically
+// set by the tcpconn.Dirsuptor, take care of not sending the re-queued, flagged packets back to the NFQHandler, and
+// instead direct them to the REJECT chain.
+// If a connection is not decided to be terminated, the packet is sent to the ACCEPT chain.
+type NFQHandler struct {
+	NFQConfig  NFQConfig
+	Disruption Disruption
+}
+
+// Handle sets up a nfqueue handler and starts handling packets sent to it. It blocks until an error occurs, or until
+// the supplied context is canceled.
+func (h NFQHandler) Handle(ctx context.Context) error {
+	nfq, err := netfilter.NewNFQueue(h.NFQConfig.QueueID, 1, netfilter.NF_DEFAULT_PACKET_SIZE)
+	if err != nil {
+		return fmt.Errorf("creating nfqueue: %w", err)
+	}
+
+	// FIXME: This complex defer works around the fact that sometimes nfq.Close() never returns.
+	defer func() {
+		closeCh := make(chan struct{})
+		go func() {
+			nfq.Close()
+			close(closeCh)
+		}()
+
+		select {
+		case <-time.After(time.Second):
+			log.Println("Time out waiting to close NFQueue")
+		case <-closeCh:
+		}
+	}()
+
+	packets := nfq.GetPackets()
+
+	// TODO: This loop can be run by multiple workers in parallel for better performance.
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case p := <-packets:
+			h.handlePacket(p)
+		}
+	}
+}
+
+// shouldTerminate returns whether the connection to which p belongs should be terminated as part of the disruption.
```

Comment:
I would separate this into a different object so it can be tested in isolation only by injecting packages). Also, this object could implement an interface (for example, `DropPolicy`) to facilitate testing and comparing different implementations.

Reply:
I'd say testing the policies is indeed a bit secondary, as right now what we have is merely probabilistic. But the separation does make sense to me, specially since there is a very loose coupling between the actual "netfilter stuff" and the termination decision (just a `[]byte`). I'll play with that.